Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.cpp

```diff
@@ -0,0 +1,132 @@
+    red_mf.setVal(0._rt);
+    ppc_mf.setVal(0._rt);
+    auto& pc = warpx.GetPartContainer().GetParticleContainer(m_ispec);
+    // Copy over compiled parser function so that it can be captured by value in the lambda
+    auto map_fn = m_map_fn;
+    ParticleToMesh(pc, red_mf, m_lev,
+            [=] AMREX_GPU_DEVICE (const WarpXParticleContainer::SuperParticleType& p,
+                amrex::Array4<amrex::Real> const& out_array,
+                amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& plo,
+                amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& dxi)
+            {
+                // Get position in WarpX convention to use in parser. Will be different from
+                // p.pos() for 1D and 2D simulations.
+                amrex::ParticleReal xw = 0, yw = 0, zw = 0;
+                get_particle_position(p, xw, yw, zw);
+
+                // Get position in AMReX convention to calculate corresponding index.
+                // Ideally this will be replaced with the AMReX NGP interpolator
+                amrex::ParticleReal x = 0, y = 0, z = 0;
+                amrex::Real lx, ly, lz;
+                int ii = 0, jj = 0, kk = 0;
+                // Always do x direction
+                x = p.pos(0);
+                lx = (x - plo[0]) * dxi[0];
+                ii = static_cast<int>(amrex::Math::floor(lx));
+                if (AMREX_SPACEDIM >= 2) {
+                    y = p.pos(1);
+                    ly = (y - plo[1]) * dxi[1];
+                    jj = static_cast<int>(amrex::Math::floor(ly));
+                }
+                if (AMREX_SPACEDIM == 3) {
+                    z = p.pos(2);
+                    lz = (z - plo[2]) * dxi[2];
+                    kk = static_cast<int>(amrex::Math::floor(lz));
+                }
+
+                // Fix dimensions since parser assumes u = gamma * v / c
+                amrex::ParticleReal ux = p.rdata(PIdx::ux) / PhysConst::c;
+                amrex::ParticleReal uy = p.rdata(PIdx::uy) / PhysConst::c;
+                amrex::ParticleReal uz = p.rdata(PIdx::uz) / PhysConst::c;
+                amrex::Real value = map_fn(xw, yw, zw, ux, uy, uz);
+                amrex::Gpu::Atomic::AddNoRet(&out_array(ii, jj, kk, 0), p.rdata(PIdx::w) * value);
+            });
+    // Add the weight for each particle -- total number of particles of this species
+    ParticleToMesh(pc, ppc_mf, m_lev,
+            [=] AMREX_GPU_DEVICE (const WarpXParticleContainer::SuperParticleType& p,
+                amrex::Array4<amrex::Real> const& out_array,
+                amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& plo,
+                amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& dxi)
+            {
+
+                // Get position in AMReX convention to calculate corresponding index.
+                // Ideally this will be replaced with the AMReX NGP interpolator
+                amrex::ParticleReal x = 0, y = 0, z = 0;
```

Comment:
Same, do these maybe need the `_rt` suffix for consistency?
```suggestion
                amrex::ParticleReal x = 0._rt, y = 0._rt, z = 0._rt;
```